Changes to Web Rope and Web Diver mutations

acssimpson opened the pull request:
#### Summary

SUMMARY: Features "Added additional utility functionality to Web Diver and Web Rope mutations to diversify the spider line, and changed the kCal limitation in the mutation_ui.cpp to align with the kCal limitation present in mutation.cpp."

#### Purpose of change

Both to diversify the spider mutation line by giving it tools that make it more suited for urban exploration, to clear up confusion about Web Diver not actually helping you descend from ledges (despite its description).

#### Describe the solution

Web Rope:
Now, when you activate Web Rope, instead of placing a rope in your inventory, it will spawn an f_rope_up_web piece of furniture, which behaves identically to the grappling hook's f_rope_up except that it turns into a long rope when taken down.

Web Diver:
Instead of only helping you descend half broken stairs, web diver will now also let you safely descend in any of the situations in which wings would would allow you to descend (namely ledges & pits). 

kCal Requirement Change:
I also changed the kCal check that stops you from using a mutation in the UI from 80% to 50%.  At least one other player has noted that 80% was high enough they thought it was a bug, and I noticed that it confusingly clashed with the exact same check that occurs slightly higher up the stack in the activate_mutation method (which is where I got the value of 50%).

#### Describe alternatives you've considered

Both of these mutations keep their existing functionality with my changes. If the new functionality is deemed too powerful when viewed in combination with the existing functionality, we could change/remove the existing functionality.

#### Testing

Web Rope: I create web rope/attempted to create a web rope across a variety of terrain types, and in all situations in which I could create a normal f_rope_up.

Web Diver:
I jumped off a bunch of different buildings, and fell in pits, verifying that the mutation prevented me from taking fall damage, and the correct message played.

Relgar commented:
GOOD GOD YES

olanti-p commented on src/iexamine.cpp:
```diff
@@ -4426,6 +4430,38 @@ void iexamine::ledge( player &p, const tripoint &examp )
+        case 3: {
+
+            if( ( p.get_kcal_percent() < 0.5f ) || ( p.get_thirst() > thirst_levels::dehydrated ) ) {
+                p.add_msg_if_player( _( "You just don't have it in you to spin that much web..." ) );
```
```suggestion
                p.add_msg_if_player( _( "You just don't have it in you to spin that much web…" ) );
```

olanti-p commented on src/trapfunc.cpp:
```diff
@@ -1077,6 +1084,9 @@ bool trapfunc::sinkhole( const tripoint &p, Creature *c, item *i )
                                    _( "You step into a sinkhole!  Throw your rope out to try to catch something?" ) ) ) {
             success = sinkhole_safety_roll( pl, itype_rope_30, 12 );
+        } else if( pl->has_trait( trait_WEB_RAPPEL ) && query_yn(
+                       _( "You step into a sinkhole!  Throw a web out to try to catch something? " ) ) ) {
```
```suggestion
                       _( "You step into a sinkhole!  Throw a web out to try to catch something?" ) ) ) {
```

olanti-p reviewed (commented):
Clang-tidy also complains about text style in C++ files.

Coolthulhu commented on src/iexamine.cpp:
```diff
@@ -4426,6 +4430,38 @@ void iexamine::ledge( player &p, const tripoint &examp )
+
+                    g->m.ter_set( dest, t_web_bridge );
+                }
+                p.mod_stored_nutr( 60 );
```
Make sure it makes you more hungry, not less. Also, would be better to use kcal directly, nutr is deprecated compatibility thing.

acssimpson replied:
ahh, okay. I actually grabbed the mod_stored_nutr from the other mutation check. I'll change both.

Kind of funny that it makes you less hungry. Whoops.

Coolthulhu commented on src/map.cpp:
```diff
@@ -2917,15 +2917,61 @@ void map::collapse_at( const tripoint &p, const bool silent, const bool was_supp
+    }
+}
+
+void map::check_for_suspension_collapse( const tripoint &point )
```
The name suggests it's just a check. Seems that you are collapsing unsupported suspensions here.

acssimpson replied:
It checks to see if its supported first. This is called on any web tile that is adjacent to a tile that has been destroyed, but that doesn't always mean that the suspended tile will be destroyed.

Coolthulhu replied:
What it does is collapsing unsupported suspension tiles. The name should reflect that.

Coolthulhu commented on src/map.cpp:
```diff
@@ -2917,15 +2917,61 @@ void map::collapse_at( const tripoint &p, const bool silent, const bool was_supp
+    if(
+        ter( point + tripoint_east ) != t_open_air
+        && ter( point + tripoint_west ) != t_open_air ) {
+        secure = true;
```
You can return early to avoid having this variable.

Would look cleaner if you had two functions: one that just checks for support and returns `bool` and another that uses this `bool` to propagate collapse.

acssimpson replied:
done :)

Coolthulhu commented on src/mapdata.cpp:
```diff
@@ -183,7 +183,8 @@ static const std::unordered_map<std::string, ter_bitflags> ter_bitflags_map = {
         { "Z_TRANSPARENT",            TFLAG_Z_TRANSPARENT },  // Doesn't block vision passing through the z-level
-        { "SUN_ROOF_ABOVE",           TFLAG_SUN_ROOF_ABOVE }   // This furniture has a "fake roof" above, that blocks sunlight (see #44421).
+        { "SUN_ROOF_ABOVE",           TFLAG_SUN_ROOF_ABOVE },  // This furniture has a "fake roof" above, that blocks sunlight (see #44421).
+        { "SUSPENDED",                TFLAG_SUSPENDED }       // This furniture is suspended between other terrain, and will cause a cascading failure on break.
```
This array is for flags important for performance. You don't need entries in here if you aren't using the bitmap flag checks (`has_flag(p, TFLAG_SUSPENDED)`). Bitmap checks are faster, but if you aren't doing 20+ of them per game second or map load, it's usually not necessary.

acssimpson replied:
will leave this as per our discussion about doing suspension checks on load.

Coolthulhu commented on data/json/furniture_and_terrain/furniture-tools.json:
```diff
@@ -1141,5 +1141,26 @@
+    "symbol": "<",
+    "color": "white",
+    "move_cost_mod": 1,
+    "required_str": 10,
```
That's for grab+move. Can you grab the rope and move it around?

acssimpson replied:
Oh. I have no idea. This entry was actually copied 1:1 from the grappling hook rope. I'll test with both I guess.

acssimpson replied:
so you totally can grab and move ropes. its pretty weird. Should we change that?

Coolthulhu replied:
Would be better to make it -1 str, so that it can't be moved around with grab

Coolthulhu commented on src/iexamine.cpp:
```diff
@@ -4426,6 +4430,38 @@ void iexamine::ledge( player &p, const tripoint &examp )
+                break;
+            }
+            const int range = 6; //this means we could web across a gap of 5.
+            int successRange = 0;
```
`snake_case` preferred for consistency.

Coolthulhu commented on src/map.cpp:
```diff
@@ -2917,15 +2917,61 @@ void map::collapse_at( const tripoint &p, const bool silent, const bool was_supp
+    //if all fail, destroy, and call check nearby suspension on self.
+    if( !secure ) {
+        ter_set( point, t_open_air );
+        furn_set( point, f_null );
```
I see 2 possible edge cases to check here.

Furniture:
* Make a web at least 2 tiles long
* Grab a bookshelf
* Push it onto the web
* Destroy the other tile of the web

Vehicle:
* As above, but cart instead of bookshelf

The first case sounds like it will completely delete the furniture, without anything left over. But web bridges generally allow building on them - they even have the "FLAT" flag.
Sounds like it would be better to restrict the suspension checks to terrain and not allow it on furniture.

The vehicle may either stay in air or fall. If it falls, it's working as expected.

acssimpson replied:
I didn't realize a bookshelf made the tile stop counting as t_open_air. interesting. I will test.

acssimpson replied:
Both tested, and it is robust to both.

Coolthulhu commented on src/map.cpp:
```diff
@@ -3372,7 +3418,13 @@ void map::bash_ter_furn( const tripoint &p, bash_params &params )
+    if( suspended ) {
+        //Its important that we change the ter value before recursing, otherwise we'll hit an infinite loop
+        //This could be prevented by aseembling a visited list, but in order to avoid that cost, we're going
+        //build our recursion to just be resilliant.
```
Typos. Should be: "assembling", "resilient".

Tiny note: most comments start with `//` and a space.

Coolthulhu commented on src/mapdata.cpp:
```diff
@@ -663,6 +664,7 @@ ter_id t_null,
        t_stairs_down, t_stairs_up, t_manhole, t_ladder_up, t_ladder_down, t_slope_down,
        t_slope_up, t_rope_up,
        t_manhole_cover,
+       t_web_bridge,
```
Those are for legacy stuff, no need to add new entries here.

acssimpson replied:
ngl I just added my entry to all of the locations in which similar tiles had entries.  I'll remove this one.

Coolthulhu commented on src/mapdata.cpp:
```diff
@@ -886,6 +888,7 @@ void set_ter_ids()
     t_water_moving_sh = ter_id( "t_water_moving_sh" );
     t_water_sh = ter_id( "t_water_sh" );
     t_water_dp = ter_id( "t_water_dp" );
+    t_web_bridge = ter_id( "t_web_bridge" );
```
Also legacy

acssimpson replied:
👍

Coolthulhu commented on src/mapdata.h:
```diff
@@ -561,6 +562,7 @@ extern ter_id t_null,
        t_stairs_down, t_stairs_up, t_manhole, t_ladder_up, t_ladder_down, t_slope_down,
        t_slope_up, t_rope_up,
        t_manhole_cover,
+       t_web_bridge,
```
Legacy here

acssimpson replied:
😦 👍

Coolthulhu commented on data/json/mutations/mutations.json:
```diff
@@ -3623,15 +3623,24 @@
+    "type": "mutation",
+    "id": "WEB_BRIDGE",
+    "name": { "str": "Arachnitect" },
+    "points": 5,
```
Should be 1 or 2 points.

Coolthulhu commented on src/iexamine.cpp:
```diff
@@ -4426,6 +4430,38 @@ void iexamine::ledge( player &p, const tripoint &examp )
+
+                    g->m.ter_set( dest, t_web_bridge );
+                }
+                p.mod_stored_kcal( -60 );
```
`kcal` and `nutr` aren't 1:1. One `nutr` is something like ~8 kcal.

Coolthulhu commented on src/map.cpp:
```diff
@@ -2917,15 +2917,63 @@ void map::collapse_at( const tripoint &p, const bool silent, const bool was_supp
+    }
+}
+
+bool map::check_suspension_validity( const tripoint &point )
```
Good `bool` function names start with `is`.

Coolthulhu commented on src/iexamine.cpp:
```diff
@@ -4426,6 +4430,38 @@ void iexamine::ledge( player &p, const tripoint &examp )
+                    tripoint dest( p.posx() + i * sgn( examp.x - p.posx() ), p.posy() + i * sgn( examp.y - p.posy() ),
+                                   p.posz() );
+
+                    g->m.ter_set( dest, t_web_bridge );
```
If you change it to `ter_id( "t_web_bridge" )`, you won't need the legacy bindings.
The most consistent way would be to declare it near top of the file as a static id.

Coolthulhu commented on src/map.cpp:
```diff
@@ -2917,15 +2917,63 @@ void map::collapse_at( const tripoint &p, const bool silent, const bool was_supp
     // that's not handled for now
 }
 
+void map::check_nearby_suspension( const tripoint &point )
```
`collapse_unsupported_suspensions` or something like that would be better.

Coolthulhu commented on src/mutation.cpp:
```diff
@@ -531,6 +532,17 @@ void Character::activate_mutation( const trait_id &mut )
     if( mut == trait_WEB_WEAVER ) {
         g->m.add_field( pos(), fd_web, 1 );
         add_msg_if_player( _( "You start spinning web with your spinnerets!" ) );
+    } else if( mut == trait_WEB_ROPE ) {
```
Could be better if you moved this functionality to attempt to go up. As in, `game::vertical_move`. This way it would be consistent with rappel down and the old, convenient rope "printing" could be preserved.

acssimpson replied:
done!

Coolthulhu commented:
As for the collapsing outside bubble, you could try to copy `map::build_floor_cache`.
That should be enough. It will do more checks than necessary, but it shouldn't be noticeably slow.

Coolthulhu commented on src/character.h:
```diff
@@ -760,6 +760,9 @@ class Character : public Creature, public visitable<Character>
         void deactivate_mutation( const trait_id &mut );
 
+        /** Removes the appropriate costs, applies mods, and recalculates sightlines (taken from activate_mutation) */
+        void deduct_mutation_cost( const trait_id &mut );
```
Two issues I see:
* The name and the comment say different things. Does it apply mutation effects, like mods, or does it just deduct the cost?
* Mutation cost is ambiguous here and could imply trait points. Mutation upkeep? Fuel mutation? Mutation use energy (not use mutation energy or mutation energy use, those would be different)?

Coolthulhu commented on src/game.cpp:
```diff
@@ -10396,8 +10398,27 @@ void game::vertical_move( int movez, bool force, bool peeking )
+                }
+
+            } else {
+                add_msg( m_info, _( "You can't climb here - you need walls and/or furniture to brace against." ) );
```
Rather than `else`, I'd have it happen in all cases where you don't spin the rope. This way you'd know that you can neither spin nor climb.

Coolthulhu commented on src/game.cpp:
```diff
@@ -10396,8 +10398,27 @@ void game::vertical_move( int movez, bool force, bool peeking )
+                    } else if( g->m.has_furn( u.pos() ) ) {
+                        add_msg( m_info, _( "There is already furniture at that location." ) );
+                    } else {
+                        add_msg( m_good, _( "You spin a climbable rope of web." ) );
```
Multiple issues here:
* You only say about the rope, but not about climbing.
* You should still require a point to attach the rope to, so that you can't build a "pillar" of unsupported ropes to the sky
* A prompt would be helpful here
* This way of climbing skips a lot of checks below. For example, grabbed cart letting go, move effects (being grabbed), moving up while at topmost z-level.

I see a nice way of solving all of them:
Have web weaving allow climbing 0 climbing cost tiles, and create a web tile when it does, but still use the rest of the code as usual.
It could go like this:
* If you can make webs, note it here. Note if you can actually spin a web here, then if you can climb without it. You could use `enum class` or two bools.
* If you can do neither, print messages and return. Otherwise keep going.
* In `point_selection_menu`, offer two sets of points:
 * One just for climbing
 * One for make rope and climb
* The set of points that isn't available could be grayed out. Alternatively, the title could say something like "Too hard to climb by hand. Can make a web rope and climb.", similar for the two other cases. 
* `point_selection_menu` will have to take and return some info on whether the point is to be reached by climbing or web-climbing. You could create a struct right above it and have it in the optional, in place of the tripoint. It only needs an extra bool for the web-ness. The struct only needs to be declared in the cpp file, since it wouldn't be used outside it.
* Right before `if( !u.move_effects( false ) ) {`, if the player picked web climb, create the web and deduct its cost.

This way, it would follow the idea of "spin web and then climb it" exactly.
If the `move_effects` failed, the next iteration wouldn't try to spin the web again, it would just try to climb the already spun one.

Coolthulhu commented on src/iexamine.cpp:
```diff
@@ -4426,6 +4432,36 @@ void iexamine::ledge( player &p, const tripoint &examp )
+            for( int i = 2; i <= range; i++ ) {
+                //break at the first non empty space encountered
+                if( g->m.ter( tripoint( p.posx() + i * sgn( examp.x - p.posx() ),
+                                        p.posy() + i * sgn( examp.y - p.posy() ), p.posz() ) ).obj().trap != tr_ledge ) {
```
Why are you checking for ledges specifically? Is there a non-open-air-terrain type you want to cover? Will this terrain type be restored properly when the bridge collapses?

Coolthulhu commented on src/map.cpp:
```diff
@@ -2917,15 +2917,63 @@ void map::collapse_at( const tripoint &p, const bool silent, const bool was_supp
+                //If suspended, make a suspension check
                 ter_set( tz, t_open_air );
                 furn_set( tz, f_null );
+                propogate_suspension_check( tz );
```
Typo: `propagate`, not `propogate`

acssimpson replied:
damnit

Coolthulhu commented on src/map.cpp:
```diff
@@ -2917,15 +2917,63 @@ void map::collapse_at( const tripoint &p, const bool silent, const bool was_supp
             // this tile used to support a roof, now it doesn't, which means there is only
             // open air above us
             if( zlevels ) {
+                //If suspended, make a suspension check
```
No need for this comment, the function name is good enough.

Coolthulhu commented on src/map.cpp:
```diff
@@ -2917,15 +2917,63 @@ void map::collapse_at( const tripoint &p, const bool silent, const bool was_supp
 
+void map::propogate_suspension_check( const tripoint &point )
+{
+    //First check all neighbors for the existence of supsension tiles.
```
Seems self-documenting enough, no need for the comment.

Coolthulhu commented on src/map.cpp:
```diff
@@ -2917,15 +2917,63 @@ void map::collapse_at( const tripoint &p, const bool silent, const bool was_supp
+
+void map::collapse_invalid_suspension( const tripoint &point )
+{
+    //if all fail, destroy, and call check nearby suspension on self.
```
Comment not needed.

Coolthulhu commented on src/map.cpp:
```diff
@@ -2917,15 +2917,63 @@ void map::collapse_at( const tripoint &p, const bool silent, const bool was_supp
+
+bool map::is_suspension_valid( const tripoint &point )
+{
+    //check the four orientations (up/down, left/right, and both diagonals)
```
Nit: comment style. Space after slashes, start with capital letter, as a sentence.
The comment should mention that those are pairs of opposing neighbors, not just 4 neighbors. But I'm not sure if it's needed.

Coolthulhu commented on src/map.cpp:
```diff
@@ -2917,15 +2917,63 @@ void map::collapse_at( const tripoint &p, const bool silent, const bool was_supp
+bool map::is_suspension_valid( const tripoint &point )
+{
+    //check the four orientations (up/down, left/right, and both diagonals)
+    if(
```
No need for this newline at start of if.

Coolthulhu replied:
Still looks weird with that empty line right after `if`

Coolthulhu commented on src/mutation.cpp:
```diff
@@ -1737,6 +1711,55 @@ bool contains_trait( std::vector<string_id<mutation_branch>> traits, const trait
+        ( mdata.fatigue && character.get_fatigue() >= fatigue_levels::exhausted ) ) {
+        // Insufficient Foo to *maintain* operation is handled in player::suffer
+        character.add_msg_if_player( m_warning, _( "You feel like using your %s would kill you!" ),
+                                     mdata.name() );
```
Sounds like this line should be extracted out of the function. The function name clearly implies just a check.
Function names being just checks is good practice, so it's better to keep the name and move the message elsewhere.

If it would result in 4+ repetitions of this message, there's another solution: have a separate function, like `can_use_mutation_warn`, which calls this one and then prints the warning if this one returns `false`.
It could be done with a `bool` parameter instead, but `bool` parameters are hard to read, since C++ doesn't have named parameters and not all IDEs provide named parameter info.

Coolthulhu commented on src/mutation.cpp:
```diff
@@ -1737,6 +1711,55 @@ bool contains_trait( std::vector<string_id<mutation_branch>> traits, const trait
     return std::find( traits.begin(), traits.end(), trait ) != traits.end();
 }
 
+bool can_use_mutation( const trait_id &mut, Character &character )
```
The `Character &` can probably be made `const Character &` here. Things should be `const`ed where possible, this conveys meaning well. And makes compiling easier.

Coolthulhu commented on src/map.cpp:
```diff
@@ -7897,6 +7952,41 @@ void map::build_floor_caches()
+    for( int smx = 0; smx < my_MAPSIZE; ++smx ) {
+        for( int smy = 0; smy < my_MAPSIZE; ++smy ) {
+            const submap *cur_submap = get_submap_at_grid( { smx, smy, zlev } );
+            const submap *below_submap = !lowest_z_lev ? get_submap_at_grid( { smx, smy, zlev - 1 } ) : nullptr;
```
You aren't using this here, only checking if it exists. Can be safely removed.

Coolthulhu commented on src/map.cpp:
```diff
@@ -7850,8 +7901,12 @@ bool map::build_floor_cache( const int zlev )
+    // We check for failed suspension here so we can piggyback off of the floor_cache's dirty logic.
+    // This allows us to only calculate suspension when a new submap is loaded.
+    // We don't nest it deeper in the floor cache loops because we want to already have established where floors are before the floorcache is built.
+    resolve_suspensions_at_level( zlev );
```
This must not be called in floor cache building. Floor cache builds floor cache, it doesn't resolve anything. Ideally, it would be `const`, the only real reason it isn't is that it gets a cache.

Coolthulhu commented:
One minor issue is that the minor changes to mutations break a really weird test for thresholds. If possible, you should also build and run tests once in a while.

You could try relaxing the `mutation_test.cpp:122` test for LUPINE category. Or adding 1-3 of kinda-related traits to the category, so that the test doesn't break.

acssimpson commented:
God clang tidy is so damn needy,.

Coolthulhu commented on src/mutation_ui.cpp:
```diff
@@ -366,8 +364,6 @@ player::power_mut_ui_result player::power_mutations_ui()
                                 ret.mut = mut_id;
                                 exit = true;
-                            } else {
-                                popup( _( "You don't have enough in you to activate your %s!" ), mut_data.name() );
```
No warning?

acssimpson replied:
Its now handled in the activation method in Mutation.cpp.

Coolthulhu commented:
I'm getting "you feel like using your arachnitect would kill you!" with every web spin.

Coolthulhu commented on src/mutation.cpp:
```diff
@@ -1737,6 +1709,57 @@ bool contains_trait( std::vector<string_id<mutation_branch>> traits, const trait
+bool can_use_mutation_warn( const trait_id &mut, const Character &character )
+{
+    const bool result = can_use_mutation( mut, character );
+    if( result ) {
```
Inverted logic here

Coolthulhu commented on src/map.cpp:
```diff
@@ -2919,13 +2921,58 @@ void map::collapse_at( const tripoint &p, const bool silent, const bool was_supp
+
+void map::collapse_invalid_suspension( const tripoint &point )
+{
+    if( !is_suspension_valid( point ) ) {
```
This causes all terrain types to collapse if a suspension check is called on them and they aren't supported from sides.
Try using debug map editor above z-level 0.

I tried this:
* Spawn 2 walls
* Spawn roofs over them
* Teleport above and try to web
* Web works
* Next turn, all terrain disappears

Coolthulhu commented on src/map.cpp:
```diff
@@ -2919,13 +2923,54 @@ void map::collapse_at( const tripoint &p, const bool silent, const bool was_supp
+void map::propagate_suspension_check( const tripoint &point )
+{
+    for( const tripoint &neighbor : points_in_radius( point, 1 ) ) {
+        if( neighbor != point && has_flag( "SUSPENDED", neighbor ) ) {
```
Can use the bitset flag overload here, for a bit more `fast`.

Coolthulhu commented on src/iexamine.cpp:
```diff
@@ -4426,6 +4432,36 @@ void iexamine::ledge( player &p, const tripoint &examp )
+
+                    g->m.ter_set( dest, t_web_bridge );
+                }
+                p.mutation_spend_resources( trait_WEB_BRDIGE );
```
I just realized you always pay 100% cost for the mutation regardless of bridge length. So 1 tile bridge will be very expensive and 5 tile bridge will be cheap.

Not necessary to be changed here, since it would require writing something like "can use mutation `n` times", but it would be good to have less flat cost in a future PR.

Coolthulhu commented:
Problem:
* Get starving
* Try to activate insect wings

Expected:
* You can see why don't they activate

Got:
* You have to exit the menu to see the message

Coolthulhu commented:
If you don't have enough food to spin a web to go upwards, you don't get the message "there is nothing to attach your web to".

Coolthulhu commented:
When climbing down a web, it says "grappling hook".

Coolthulhu commented:
So far, out of those problems, only the lack of popup is serious enough to hold the PR up, because it's a regression that affects many mutations.
Should be ez fix, though.

acssimpson commented:
> If you don't have enough food to spin a web to go upwards, you don't get the message "there is nothing to attach your web to".

Fixed!



> Problem:
> 
> * Get starving
> * Try to activate insect wings
> 
> Expected:
> 
> * You can see why don't they activate
> 
> Got:
> 
> * You have to exit the menu to see the message

Fixed!!
